DMP-3345 Fixing concurrent daily list processing

davet1985 opened the pull request:
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:


### JIRA link (if applicable) ###



### Change description ###

**Problem:**
The daily list can be trigger using the POST /dailylists/run endpoint. This endpoint is useful for running daily lists for a specific listing courthouse and is (or will be) locked down to SUPER_ADMIN only.

**Solution:**
Utilise the automated tasks locking mechanism when running the daily list processing, to ensure that it's not already locked by the automated task running on schedule or another manual execution.
- the results in calls to POST /dailylists/run getting a 409 response if the task is already locked, preventing concurrent executions.

**Implementation:**
Exposing some methods from the automated task package through an "api" interface.
Adding a LockService, exposed within the "api" package.
- used by the daily list controller to check if the task is locked
- used by the daily list processor to execute the task with a lock

Using the new LockService within the automated tasks
- to get the locking task executor
- to get the default lock times

**Additionally:**
- updating DarNotifyError constructor to call super
- using ConcurrentMap in SystemUserHelper - to fix a ConcurrentModificationException that was seen in the logs

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

chrisbellingham-hmcts commented on src/main/java/uk/gov/hmcts/darts/task/service/impl/LockServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+@RequiredArgsConstructor
+public class LockServiceImpl implements LockService {
+
+    private static final int DEFAULT_LOCK_AT_MOST_SECONDS = 600;
+    private static final int DEFAULT_LOCK_AT_LEAST_SECONDS = 20;
```
Consider declaring these as `Duration` to avoid the need for the descriptive name and convertion in the getters:

```suggestion
    private static final Duration DEFAULT_LOCK_AT_MOST = Duration.ofSeconds(600);
    private static final Duration DEFAULT_LOCK_AT_LEAST = Duration.ofSeconds(20);
```

chrisbellingham-hmcts commented on src/main/java/uk/gov/hmcts/darts/dailylist/controller/DailyListController.java:
```diff
@@ -95,12 +97,12 @@ public ResponseEntity<Void> dailylistsHousekeepingPost() {
+        var taskName = PROCESS_DAILY_LIST_TASK_NAME.getTaskName();
+        Optional<AutomatedTaskEntity> automatedTaskEntity = automatedTasksApi.getTaskByName(taskName);
+        if (automatedTaskEntity.isPresent() && automatedTasksApi.isLocked(automatedTaskEntity.get())) {
+            throw new DartsApiException(DAILY_LIST_ALREADY_PROCESSING);
         }
-
+        CompletableFuture.runAsync(() -> processor.processAllDailyListsWithLock(listingCourthouse));
```
Is there an opportunity for a race condition if the automated task takes the lock between `automatedTasksApi.isLocked()` and `processor.processAllDailyListsWithLock()`? Unlikely I suppose, but what would happen?

davet1985 replied:
Yeah I guess there is a slight chance of it happening. To be honest the chances are so slim and if it did happen then no harm would be done, just the Daily list running twice.

davet1985 commented:
@chrisbellingham-hmcts I've rebased this and fixed up conflicts. It's been a while since your first review, but would you mind giving it another review please?
